File: leetcode-104.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type TreeNodeList struct {
	Node  *TreeNode
	Depth int
}

func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	leftHeight := maxDepth(root.Left)
	rightHeight := maxDepth(root.Right)
	return max(leftHeight, rightHeight) + 1
}

func max(x, y int) int {
	if x < y {
		return y
	}
	return x
}
func dfsMaxDepth(root *TreeNode) int {
	var nodeList []TreeNodeList
	if root != nil {
		newTreeNodeList := TreeNodeList{Node: root, Depth: 1}
		nodeList = append(nodeList, newTreeNodeList)
	}
	depth := 0
	for len(nodeList) != 0 {
		newTreeNodeList := nodeList[len(nodeList)-1]
		nodeList = nodeList[:len(nodeList)-1]
		root = newTreeNodeList.Node
		currentDepth := newTreeNodeList.Depth
		if root != nil {
			depth = max(depth, currentDepth)
			newTreeNodeList := TreeNodeList{Node: root.Left, Depth: currentDepth + 1}
			nodeList = append(nodeList, newTreeNodeList)
			newTreeNodeList = TreeNodeList{Node: root.Right, Depth: currentDepth + 1}
			nodeList = append(nodeList, newTreeNodeList)
		}
	}
	return depth
}

File: leetcode-104.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if root is None:
            return 0
        left_height = self.maxDepth(root.left)
        right_height = self.maxDepth(root.right)
        return max(left_height, right_height) + 1

    def dfs_max_depth(self, root: TreeNode) -> int:
        node_list = []
        if root:
            node_list.append((1, root))
        depth = 0
        while node_list:
            current_depth, node = node_list.pop()
            if node:
                depth = max(depth, current_depth)
                node_list.append((current_depth + 1, node.left))
                node_list.append((current_depth + 1, node.right))
        return depth
